WMF Framework/Event Platform Client/EventPlatformClientLibrary.swift

```diff
@@ -0,0 +1,776 @@
+     *
+     * Regarding `$schema`: the instrumentation needs to specify which schema (and
+     * specifically which version of that schema) it conforms to. Analytics
+     * schemas can be found in the jsonschema directory of
+     * [secondary repo](https://gerrit.wikimedia.org/g/schemas/event/secondary/)
+     *
+     * As an example, if instrumenting client-side error logging, a possible
+     * `$schema` would be `/mediawiki/client/error/1.0.0`. For the most part, the
+     * `$schema` will start with `/analytics`, since there's where
+     * analytics-related schemas are collected. An example call:
+     *
+     * ```
+     * EPC.shared?.submit( "test.instrumentation", [ "$schema": "/analytics/test/1.0.0" as NSCoding ] )
+     * ```
+     *
+     * Regarding `domain`: this is *optional* and should be used when event needs
+     * to be attrributed to a particular wiki (Wikidata, Wikimedia Commons, a
+     * specific edition of Wikipedia, etc.). If the language is NOT relevant in
+     * the context, `domain` can be safely omitted. Using "domain" rather than
+     * "language" is consistent with the other platforms and allows for the
+     * possibility of setting a non-Wikipedia domain like "commons.wikimedia.org"
+     * and "wikidata.org" for multimedia/metadata-related in-app analytics.
+     * Instrumentation code should use the `host` property of a `URL` as the value
+     * for this parameter.
+     *
+     * Cases where instrumentation would set a `domain`:
+     * - reading or editing an article
+     * - managing watchlist
+     * - interacting with feed
+     * - searching
+     *
+     * Cases where it might not be necessary for the instrument to set a `domain`:
+     * - changing settings
+     * - managing reading lists
+     * - navigating map of nearby articles
+     * - multi-lingual features like Suggested Edits
+     * - marking session start/end; in which case schema and `data` should have a
+     *   `languages` field where user's list of languages can be stored, although
+     *   it might make sense to set it to the domain associated with the user's
+     *   1st preferred language – in which case use
+     *   `MWKLanguageLinkController.sharedInstance().appLanguage.siteURL().host`
+     */
+    @objc public func submit(stream: String, data: [String: NSCoding], domain: String? = nil) -> Void {
+        guard self.sharingUsageData else {
+            return
+        }
+
+        /*
+         * EventGate needs to know which version of the schema to validate against
+         */
+        let schemaKey = "$schema"
+        guard data.keys.contains(schemaKey) else {
+            DDLogError("EPC: Event data is missing the required '$schema' field")
+            return
+        }
+
+        let metaKey = "meta"
+        var meta: [String: NSCoding] = data[metaKey] as? [String: NSCoding] ?? [:]
+
+        if let domain = domain {
+            meta["domain"] = domain as NSCoding
```

Comment:
Yeah, at the time I thought I was limited to `NSCoding` to persist items in Core Data, though looking again now I can't remember how I arrived at that. The swiftier way for serialization in general is to use `Codable`, and we may be able to use generics (shoutout to @joewalsh for inspiration) for the event's data, which could potentially clean a lot of this dictionary construction and the log call site up. But I don't have any more concrete suggestions beyond that without tinkering with it myself first and seeing what works. @bearloga feel free to run with that or wait on this part till I have time to revisit (probably tomorrow afternoon) and try to arrive at more specific feedback.

Reply:
I'm like 99% sure the `NSCoding` is for Core Data storage reasons, which we can probably change to `Any` for v1 since we got rid of the persistence facilities. @tonisevener do you concur?

Reply:
Just changed `NSCoding` to `Codable` in `submit` method and we have our answer:

```
Method cannot be marked @objc because the type of the parameter 2 cannot be represented in Objective-C
```

So it looks like if we want to be able to use this library in any of the existing instrumentation written in Obj-C we have to use `NSCoding`